### ISSUE TITLE
position problem

### ISSUE THREAD

**Issue by doowon** [Wox-launcher/Wox#271]
When I am using dual monitors, the position of wox is the center of the two monitors. In other words, the wox should be the center of a specifiic monitor where the mouse is focused, but the current version doesn't do that.

**Comment by gugahoi** [comment]
Also getting the same problem even on 1.2 beta where it seems it was supposed to be fixed. I currently run 3 monitors and wox is never in the same place and it always varies based on which screen is selected when the hotkey is pressed.

**Comment by qianlifeng** [comment]
@gugahoi this is by design that Wox will always show based on which screen is selected.

**Comment by gugahoi** [comment]
@qianlifeng I understand however when I mean never in the same place I really mean never in the same place and it is most definitely not remembering positions for each screen, most of the time it seems like it gets triggered off screen for some reason too.

**Comment by bao-qian** [comment]
@gugahoi 
excuse for our bad English, I don't get you.

**Comment by highpowerxh** [comment]
Actually, I also have the same problem on my new Surface Pro 4. Everything is working correctly on my workstation at work. However on my surface, due to its high PPI, I have the display setting magnified to 200%, and this will make the wox (default position x,y) showing at (2x,2y) from the top left. Whenever I set the window to be in the center of the screen, the position will be moved towards top left after restart of the app.
在我公司的电脑上一切正常,但是我的surface pro 4上就有问题, 因为PPI比较高, 所以显示设置默认所有的文字和app是放大到200%的, 假设Wox的默认启动位置是(x,y),那么在我电脑上启动位置就大概是(2x,2y),我没具体测试改回100%会如何.
然后就是即使我设置了保存上次的窗口位置,重启应用之后窗口的位置会向左上移动.
另外我已经把Wox放到了不需要管理员权限的目录, 我发现重启应用依旧会报错. 我找到的原因是已经启动的everything并没有被关闭,如果手动退掉everything就不会报错了.

**Comment by lances101** [comment]
@highpowerxh, PR #370 seems to fix this. Tried setting up scaling to 200% on Windows 10 and Wox seems to handle the position just fine.  Also, what version were you using?
Could you [try this build](https://ci.appveyor.com/project/qianlifeng/wox/build/1.2.0.517/artifacts) that includes the changes from it? (you need the [Output/release-binary.zip](https://ci.appveyor.com/api/buildjobs/vy425ubsx3mq0txh/artifacts/Output/release-binary.zip))

@happlebao, regarding @gugahoi's comment,what he means is that ideally Wox should store a position for each screen (and at the date when he wrote it, even storing one position was buggy).

**Comment by highpowerxh** [comment]
@lances101 I'm on the latest build, but in case there was a regression, tried 370 version and here's the screenshot![image](https://cloud.githubusercontent.com/assets/1772748/11314321/545bb846-8f99-11e5-95bb-0d521b8cd50d.png)

**Comment by highpowerxh** [comment]
@lances101 did you sign out after bumping to 200%, wox is one of the those tho "won't response to scaling until you signed out"
And I tried changing to 100% and wox shown at correct position. back to the screenshot case after change back to 200%.
Sorry for keep editing frequently. hope this will not spam your inbox, lol. I'll try looking into the code if I have spare time.
Thanks

**Comment by highpowerxh** [comment]
Let me show you what will happen if the remember last launch position option is checked.
Place the bar at a perfect position, exit (not hide wox)
![2015-11-20 1](https://cloud.githubusercontent.com/assets/1772748/11314652/60773c06-8f9c-11e5-9f99-4d0c8dc8dae2.png)
restart #1:
![2015-11-20 3](https://cloud.githubusercontent.com/assets/1772748/11314653/6077b3f2-8f9c-11e5-8258-d6f49c1b8cbc.png)
restart #2:
![2015-11-20 4](https://cloud.githubusercontent.com/assets/1772748/11314650/607637f2-8f9c-11e5-8051-c37b0290d9a6.png)
restart #3:
![2015-11-20 5](https://cloud.githubusercontent.com/assets/1772748/11314651/60763e0a-8f9c-11e5-8165-fc84ec989fb2.png)
So, basically it is keep moving towards the top left corner. My guess would be (x/2,y/2) each restart

**Comment by lances101** [comment]
@highpowerxh editing does not spam the inbox, only new comments actually send an email, so feel free :smiley: 

Yep, Windows did not prompt me to log out, so I did not. I am now trying to replicate your error after logging out by changing the position and restarting Wox (without hiding). 

For now I can only confirm that it does exist on the [latest master-branch build (1.2.0.562)](https://ci.appveyor.com/project/qianlifeng/wox/build/1.2.0.562), but sadly I cannot replicate it on the [PR #370 build (1.2.0.517)](https://ci.appveyor.com/project/qianlifeng/wox/build/1.2.0.517)

I will look into it and add some DPI-related logic to the whole positioning thing. Right now it sticks to pixels... Which are not the most reliable of things. 

Also, on a unrelated matter, could you please try disabling "Remember last location" and see if Wox is actually in the center of the screen? Thanks. **Edit**, forget about this, you do get the bug in a screenshot you provided.

**Comment by DarrenCodes** [comment]
Could everyone experiencing this problem try this [build](https://github.com/DarrenCodes/Wox/releases/download/bugfix/Relative.Positioning.Fix.zip).

If your display is scaled, please enter the scale percentage in the Settings window of Wox in the appropriate textbox.

I would appreciate the feedback.

**Comment by highpowerxh** [comment]
Once the scale percentage is set, everything works, both remember last position or not. Just wondering if there's a way to get the percentage with out user's input
anyways, thx for the fix

**Comment by lances101** [comment]
@highpowerxh , [this build should detect](https://github.com/lances101/Wox/releases/download/bugfix-411/bugfix-411-build-1.zip) the DPI and convert it into **DIP** (device independent pixels). Cannot test it myself since I don't have access to a high DPI screen.

**Comment by highpowerxh** [comment]
@lances101 this build shows the bar at the correct position. However, it crashes shortly after startup

**Comment by lances101** [comment]
The crash is kind of intended (stripped everything possible for the size, as long as it launches). 
I'll free my schedule by tomorrow evening and will do a PR out of the version that you tried.

**Comment by highpowerxh** [comment]
@lances101 cool, thx

**Comment by DarrenCodes** [comment]
@highpowerxh I'm glad I could help. There isn't a reliable way of detecting the scale factor that I'm aware of, the build I made is a temporary fix. 
But @lances101 is trying something different, so maybe that's the route to go for Wox.

I think the key parts of the positioning of the Wox window needs to be addressed in order to have the fix accepted. From what I'm aware of, they are:
1. Window is always centered when "Remember launch ..." is not selected.
2. Window maintains position when moving across multiple screens whether "Remember launch ..." is selected or not.
3. Window maintains position when Window is hidden and then displayed.
4. Window position is saved correctly for use upon relaunch when "Remember launch ..." is selected.

My build passed these requirements on my PC. It needs to be tested more across different devices though. @highpowerxh Could you confirm what requirements passed on your device for any builds that you tested in this regard? It helps identify what still needs to be worked on.

Thanks for the feedback.

**Comment by bao-qian** [comment]
Crash is due to the SSL certificate expired. Merge the latest changes from master should fix this.

**Comment by highpowerxh** [comment]
@DarrenCodes The concern about the user input scale percentage is actually the _#_2 in the checklist.
Because the fact is that different monitor can have different scale percentage.
I tested in two secarios:
1. Connect to an external monitor[^1] while Wox is running. Wox show and hide correctly on SP4[^2], flicker on the and disappear on external monitor
2. Connect to an external monitor while Wox is closed, and then start Wox. Wox crashed immediately.

[^1]: External Monitor 1920*1080
[^2]: Surface Pro4 with Windows 10 Enterprise

P.S. _#_1 _#_3 _#_4 passed on single monitor

**Comment by DarrenCodes** [comment]
@highpowerxh Thanks.

The scale percentage pertains to this scale setting in Windows' control panel:
![](https://cloud.githubusercontent.com/assets/13368437/11744868/80824c76-a01c-11e5-81a7-794d8815ff01.jpg?raw=true)
This setting applies to all connected screens.

Please try this [build](https://github.com/DarrenCodes/Wox/releases/download/bugfix/Relative.Positioning.Fix.v1.0.1.zip) and let me know if the problems persist.

**Comment by highpowerxh** [comment]
Hi @DarrenCodes, from your screenshot I can see it's windows 7
The support for high DPI screen has been significantly improved on windows 10
Right now it is scaling setting per monitor
![scaling setting](http://www.dotnetmafia.com/blogs/dotnettipoftheday/W10B10074DisplayScalingSurfacePro3_669BCFD0.png)

**Comment by DarrenCodes** [comment]
@highpowerxh Ah, I see. 

DIP would be the best approach most probably.
The rest of the team will have to decide on what to do in this regard.

Having Wox remember the position it is placed on on each screen is possible, but DIP would be more flexible, and provide a more predictable experience, which makes things easier for the user.

For the time being, would you perhaps want the Wox Window to be fixed to one screen rather than follow the cursor? I think I can do that without much effort.

**Comment by highpowerxh** [comment]
@DarrenCodes It does not bother me too much now. I don't use secondary monitor with SP4 very often and the build I use on my desktop (1080P Monitors) works fine. I'll just wait until this problem is completely resolved.
Thanks for keeping tracking this bug.

**Comment by lances101** [comment]
@DarrenCodes At first I was considering converting to DIP myself, by grabbing the DPI from the `Graphics` class and then calculating everything inside of Wox. Turns out `PresentationSource` allows converting the values respective to the device where the form is located, which from my point of view is "safer" than implementing that ourselves. 

@highpowerxh I've created a new PR (#423) and AppVeyor has already made a fresh build (grab the [release-binary.zip](https://ci.appveyor.com/project/qianlifeng/wox/build/1.2.612/artifacts)). This PR also fixes Wox's notifications. Try hitting Shift+Enter and selecting "Set as topmost", you should get a notification in your lower right.

**Comment by highpowerxh** [comment]
@lances101 Hi, I tested and its working, but I noticed that the bar will flicker, like show>disappear>then show persistently.  And I think you're missing one Null check for the "set as topmost" option. If the input is empty, shift+enter will throw an error

**Comment by lances101** [comment]
The flickering is not part of the positioning problem... Please create a new issue with detail of when it happens 

The shift+enter menu crash, is a bug, I've created an issue for it (#434)

**Comment by bao-qian** [comment]
@highpowerxh Hi, please add details to #435 for the flickering problem
1. your installed .net version, os version
2. Wox version/build

for .net version
Download this：http://www.asoft.be/prod_netver
And paste info like this to #435 :

```
<32Bit>
2.0.50727.8669
  ->C:\Windows\Microsoft.NET\Framework\v2.0.50727
4.6.1055.0
  ->C:\Windows\Microsoft.NET\Framework\v4.0.30319

<64Bit>
2.0.50727.8669
  ->C:\Windows\Microsoft.NET\Framework64\v2.0.50727
4.6.1055.0
  ->C:\Windows\Microsoft.NET\Framework64\v4.0.30319

< Installed .NET Frameworks >
.NET FW 2.0 SP 2  (CLR:2.0)
.NET FW 3.0 SP 2  (CLR:2.0)
.NET FW 3.5 SP 1  (CLR:2.0)
.NET FW 4.6 Unknown build:394271 (CLR:4.0)
```

**Comment by highpowerxh** [comment]
@happlebao Done, BTW, the tool download page is http://www.asoft.be/prod_netver.html

**Comment by bao-qian** [comment]
@highpowerxh thanks.

**Comment by lances101** [comment]
So, does using **Device Independent Pixels** fix the positioning problem?

**Comment by bao-qian** [comment]
@Doowon @gugahoi @highpowerxh @DarrenCodes 
please confirm this is resolved by using [build 1.2.612](https://ci.appveyor.com/project/qianlifeng/wox/build/1.2.612/artifacts).

@lqzhgood  #439 
请用 [build 1.2.612](https://ci.appveyor.com/project/qianlifeng/wox/build/1.2.612/artifacts) 来确认问题是否已解决。

@koobs #449 
please confirm this is resolved by using [build 1.2.612](https://ci.appveyor.com/project/qianlifeng/wox/build/1.2.612/artifacts).

**Comment by jondcoleman** [comment]
How do I apply the update to an existing installation?  I'm kind of a n00b with windows applications like this.

**Comment by bao-qian** [comment]
@jondcoleman sorry for delay, just unzip the new folder to the existing location.

**Comment by zhaoguoyuan** [comment]
Just tested with build 1.2.612, the problem get solved on my pc.Thanks Man!

**Comment by bao-qian** [comment]
@zhaoguoyuan  thanks for report.

**Comment by specter119** [comment]
@bao-qian 

this issue still exists for me:

wox version 1.3.524

%AppData%\Wox\Settings\Settings.json

```json
  "WindowLeft": 1114.0,
  "WindowTop": 4.0,
  "RememberLastLaunchLocation": true,
```

but Wox will still locates in the middle of the window every time it starts.
